### PR TITLE
Do not bump clfs on seq mismatch when before stream LastSeq

### LINKED ISSUE #5205: Nats server pod gets stuck in KV stream catchup after restart
### Observed behavior

I discovered this when exploring key value and testing resillience and performace among other things.

I used a helm install in kubernetes.

While updating a key value bucket (in memory, with replication 3) with about 1000 updates/s I restarted one pod. It never came back up and it got stuck in a stream catchup state (waited over 20 minutes). Subsequent restarts did not resolve the issue. It only got resolved after the stream was deleted (or in subsequent tests when all nats pods where stopped).

See also this thread in slack: https://natsio.slack.com/archives/C06EN6HCWE4/p1710151450266159

Stream:
```
Information for Stream KV_gamesession-state created 2024-03-06 16:42:33

              Subjects: $KV.gamesession-state.>
              Replicas: 3
               Storage: Memory

Options:

             Retention: Limits
       Acknowledgments: true
        Discard Policy: Old
      Duplicate Window: 2m0s
            Direct Get: true
     Allows Msg Delete: false
          Allows Purge: true
        Allows Rollups: true

Limits:

      Maximum Messages: unlimited
   Maximum Per Subject: 1
         Maximum Bytes: unlimited
           Maximum Age: 30m0s
  Maximum Message Size: unlimited
     Maximum Consumers: unlimited

Cluster Information:

                  Name: nats
                Leader: nats-0
               Replica: nats-1, current, seen 97ms ago
               Replica: nats-2, current, seen 97ms ago

State:

              Messages: 10
                 Bytes: 1.7 KiB
        First Sequence: 1,245,135 @ 2024-03-11 11:10:15 UTC
         Last Sequence: 1,245,155 @ 2024-03-11 11:10:15 UTC
      Deleted Messages: 11
      Active Consumers: 0
    Number of Subjects: 10
```

Log snippet:
```
[57] 2024/03/11 10:08:28.171728 [WRN] Catchup for stream '$G > KV_gamesession-state' stalled
[57] 2024/03/11 10:08:36.806848 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:08:46.806154 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:08:56.805914 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:08:58.172655 [WRN] Catchup for stream '$G > KV_gamesession-state' stalled
[57] 2024/03/11 10:08:58.177889 [WRN] Error applying entries to '$G > KV_gamesession-state': first sequence mismatch
[57] 2024/03/11 10:08:58.178309 [WRN] Resetting stream cluster state for '$G > KV_gamesession-state'
[57] 2024/03/11 10:08:58.301144 [WRN] Catchup for stream '$G > KV_gamesession-state' resetting first sequence: 1228966 on catchup request
[57] 2024/03/11 10:09:06.806746 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:09:16.805821 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:09:26.806109 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:09:28.301893 [WRN] Catchup for stream '$G > KV_gamesession-state' stalled
[57] 2024/03/11 10:09:36.806382 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:09:46.806260 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:09:56.806782 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:09:58.302241 [WRN] Catchup for stream '$G > KV_gamesession-state' stalled
[57] 2024/03/11 10:09:58.311134 [WRN] Error applying entries to '$G > KV_gamesession-state': first sequence mismatch
[57] 2024/03/11 10:09:58.311752 [WRN] Resetting stream cluster state for '$G > KV_gamesession-state'
[57] 2024/03/11 10:09:58.650302 [WRN] Catchup for stream '$G > KV_gamesession-state' resetting first sequence: 1243279 on catchup request
[57] 2024/03/11 10:10:06.806367 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:10:16.806423 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:10:26.805642 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:10:28.652230 [WRN] Catchup for stream '$G > KV_gamesession-state' stalled
[57] 2024/03/11 10:10:36.806734 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:10:46.805759 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:10:56.806693 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:10:58.652829 [WRN] Catchup for stream '$G > KV_gamesession-state' stalled
[57] 2024/03/11 10:10:58.656426 [WRN] Error applying entries to '$G > KV_gamesession-state': first sequence mismatch
[57] 2024/03/11 10:10:58.656974 [WRN] Resetting stream cluster state for '$G > KV_gamesession-state'
[57] 2024/03/11 10:10:58.988587 [WRN] Catchup for stream '$G > KV_gamesession-state' resetting first sequence: 1243279 on catchup request
[57] 2024/03/11 10:11:06.806418 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:11:16.805928 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:11:26.806630 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:11:28.989443 [WRN] Catchup for stream '$G > KV_gamesession-state' stalled
[57] 2024/03/11 10:11:36.806795 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:11:46.806430 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
[57] 2024/03/11 10:11:56.806165 [WRN] Healthcheck failed: "JetStream stream '$G > KV_gamesession-state' is not current"
```

### Expected behavior

I expected the restarted pod to catch up and come back into the cluster.

### Server and client version

Server version: 2.10.11
Client version (golang client): v1.32.0

### Host environment

Both server and client was running in managed Kubernetes in Google Cloud, using Google Filestore as perstistence layer.

### Steps to reproduce

It is a bit hard to reproduce, I only manged to reproduce it about 1/5 tries.

What I did was:
- Run a pod with a nats client that does updates in a keyvalue as fast as possibile. This is done with 100 goroutines in paralell. The goroutines will update a set of 10 keys (so 10 goroutines for each key).
- Seems to be running for a whle (I never reproduced when running it shorter than 5 minutes).
- Restart one pod, observe it not coming up (fail ready check) and see log like snippet above.

I also tried to reproduce using a file store for the stream, but didn't manage to do it within 10 tries or so.

### PR DESCRIPTION
**Pull request by wallyqs**
Sometimes during a stream catchup after a restart, when applying entries the clfs could have been bumped due to msg not found errors from likely deleted messages, causing a stream replica to remain out of sync.

Fixes #5205

### CODE REVIEW

**Review by derekcollison — COMMENTED** [review]
I think this can cause false negatives..

**Comment by wallyqs** [comment]
rebased and added the related tests for this one, it is ready to merge now

**Review by derekcollison — APPROVED** [review]
LGTM

File: server/jetstream_cluster_4_test.go [hunk 1]
```diff
@@ -2676,3 +2677,396 @@ func TestJetStreamClusterKeyValueDesyncAfterHardKill(t *testing.T) {
...
+func TestJetStreamClusterKeyValueLastSeqMismatch(t *testing.T) {
+	c := createJetStreamClusterExplicit(t, "R3S", 3)
+	defer c.shutdown()
+
+	nc, js := jsClientConnect(t, c.randomServer())
+	defer nc.Close()
+
+	kv, err := js.CreateKeyValue(&nats.KeyValueConfig{
+		Bucket:   "mismatch",
+		Replicas: 3,
```
Comment (cbrewster) [inline server/jetstream_cluster_4_test.go:3051]: If replicas is set to `1`, this test fails:
```
--- a/server/jetstream_cluster_4_test.go
+++ b/server/jetstream_cluster_4_test.go
@@ -3048,7 +3048,7 @@ func TestJetStreamClusterKeyValueLastSeqMismatch(t *testing.T) {

        kv, err := js.CreateKeyValue(&nats.KeyValueConfig{
                Bucket:   "mismatch",
-               Replicas: 3,
+               Replicas: 1,
        })
        require_NoError(t, err)
```
```
1 lastseq 2
    jetstream_cluster_4_test.go:3070: require error, but got none
--- FAIL: TestJetStreamClusterKeyValueLastSeqMismatch (0.52s)
```

Is that expected?

Comment (derekcollison) [reply server/jetstream_cluster_4_test.go:3051]: This is a cluster specific test. So needs to be R3.

Comment (MauriceVanVeen) [reply server/jetstream_cluster_4_test.go:3051]: Have created an issue and will have a fix for this shortly: https://github.com/nats-io/nats-server/issues/5840

Comment (cbrewster) [reply server/jetstream_cluster_4_test.go:3051]: > This is a cluster specific test. So needs to be R3.

Makes sense, but still the observed behavior of the KV itself is different here depending on whether or not the KV happens to be clustered. All the ops are mutations too which should all go to the leader, so I'd expect R1 and R3 to behave the same in this case. Not suggesting that we change this test to R1, but that the behavior here should be consistent for both.

**Pull request merged**